system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+ ****************************************************************************/
+
+static int      sanity_check(FAR char *str);
+static uint32_t hash_calc(void);
+static int      get_setting(FAR char *key, FAR setting_t **setting);
+static size_t   get_string(FAR setting_t *setting, FAR char *buffer,
+                         size_t size);
+static int      set_string(FAR setting_t *setting, FAR char *str);
+static int      get_int(FAR setting_t *setting, FAR int *i);
+static int      set_int(FAR setting_t *setting, int i);
+static int      get_byte(FAR setting_t *setting, FAR int *i);
+static int      set_byte(FAR setting_t *setting, int i);
+static int      get_bool(FAR setting_t *setting, FAR int *i);
+static int      set_bool(FAR setting_t *setting, int i);
+static int      get_float(FAR setting_t *setting, FAR double *f);
+static int      set_float(FAR setting_t *setting, double f);
+static int      get_ip(FAR setting_t *setting, struct in_addr *ip);
+static int      set_ip(FAR setting_t *setting, struct in_addr *ip);
+static int      load(void);
+static int      save(void);
+static void     signotify(void);
+void            dump_cache(union sigval ptr);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct
+{
+  pthread_mutex_t   mtx;
+  uint32_t          hash;
+  bool              wrpend;
+  bool              initialized;
+  storage_t         store[CONFIG_SYSTEM_SETTINGS_MAX_STORAGES];
+  struct notify_s   notify[CONFIG_SYSTEM_SETTINGS_MAX_SIGNALS];
+#if defined(CONFIG_SYSTEM_SETTINGS_CACHED_SAVES)
+  struct sigevent   sev;
+  struct itimerspec trigger;
+  timer_t           timerid;
+#endif
+} g_settings;
+
+setting_t map[CONFIG_SYSTEM_SETTINGS_MAP_SIZE];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: hash_calc
+ *
+ * Description:
+ *    Gets a setting for a string value
+ *
+ * Input Parameters:
+ *    none
+ * Returned Value:
+ *   crc32 hash of all the settings
+ *
+ ****************************************************************************/
+
+static uint32_t hash_calc(void)
+{
+  return crc32((FAR uint8_t *)map, sizeof(map));
+}
+
+/****************************************************************************
+ * Name: get_setting
+ *
+ * Description:
+ *    Gets a setting for a string value
+ *
+ * Input Parameters:
+ *    key        - key of the required setting
+ *    setting    - pointer to pointer for the setting
+ *
+ * Returned Value:
+ *   The value of the setting for the given key
+ *
+ ****************************************************************************/
+
+static int get_setting(FAR char *key, FAR setting_t **setting)
+{
+  DEBUGASSERT(*setting == NULL);
```

Comment:
This is a private function not public, so isn't DEBUGASSERT correct here? It's the same with the other private functions or have I missed the point?

Reply:
I don't think that a "private" function is to be treated differently, as it will be essentially be called from a "public" function eventually.

Personally, I use the standard `assert()` for all userland apps, and NuttX specific macros for NuttX itself. Usually my goal is: would this application build and run successfully on Linux? (Unaltered)

But I am not sure of the conventions used in the rest of the apps. Someone else will have to enlighten us.

Reply:
I am not sure either so would like the opinions of others. All the public functions here use DEBUGASSERT so I think the qeustion is "can we reasonably recover if we don't assert". In the case of settings_iterate for example, which is a public function, if setting is NULL we subsequently memset it to 0, which sounds like a BAD IDEA?

In the case of get_setting, a private function, it is not called from user space but from another, albeit public, function. That to me sounds like DEBUGASSERT is correct here, as long as the public functions that call it are suitable protected with the correct assert type?

Once others have commented I can go through and apply that, or any other "test" to the functions - public or otherwise - and amend.